docker.compose.image.swarm.yml

```diff
@@ -0,0 +1,108 @@
+version: "3.7"
+services:
+  db_gtm_1:
+    environment:
+      - PG_HOST=0.0.0.0
+      - PG_NODE=gtm_1
+      - PG_PORT=6666
+    image: pavouk0/postgres-xl:latest
+    command: docker-cmd-gtm
+    entrypoint: docker-entrypoint-gtm
+    volumes:
+      - db_gtm_1:/var/lib/postgresql
+    networks:
+      - db
+    #    healthcheck:
+    #      test: ["CMD", "docker-healthcheck-gtm"]
+    deploy:
+      mode: global
+  db_coord_1:
+    environment:
+      - PG_GTM_HOST=db_gtm_1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=coord_1
+      - PG_PORT=5432
+    image: pavouk0/postgres-xl:latest
+    command: docker-cmd-coord
+    entrypoint: docker-entrypoint-coord
+    volumes:
+      - db_coord_1:/var/lib/postgresql
+    depends_on:
+      - db_gtm_1
+    networks:
+      - db
+    #    healthcheck:
+    #      test: ["CMD", "docker-healthcheck-coord"]
+    deploy:
+      mode: global
+  db_coord_2:
+    environment:
+      - PG_GTM_HOST=db_gtm_1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=coord_2
+      - PG_PORT=5432
+    image: pavouk0/postgres-xl:latest
+    command: docker-cmd-coord
+    entrypoint: docker-entrypoint-coord
+    volumes:
+      - db_coord_2:/var/lib/postgresql
+    depends_on:
+      - db_gtm_1
+    networks:
+      - db
+    #    healthcheck:
+    #      test: ["CMD", "docker-healthcheck-coord"]
+    deploy:
+      mode: global
+  db_data_1:
+    environment:
+      - PG_GTM_HOST=db_gtm_1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=data_1
+      - PG_PORT=5432
+    image: pavouk0/postgres-xl:latest
+    command: docker-cmd-data
+    entrypoint: docker-entrypoint-data
+    depends_on:
+      - db_gtm_1
+    volumes:
+      - db_data_1:/var/lib/postgresql
+    networks:
+      - db
+    #    healthcheck:
+    #      test: ["CMD", "docker-healthcheck-data"]
+    deploy:
+      mode: global
+  db_data_2:
+    environment:
+      - PG_GTM_HOST=db_gtm_1
+      - PG_GTM_PORT=6666
+      - PG_HOST=0.0.0.0
+      - PG_NODE=data_2
+      - PG_PORT=5432
+    image: pavouk0/postgres-xl:latest
+    command: docker-cmd-data
+    entrypoint: docker-entrypoint-data
+    depends_on:
+      - db_gtm_1
+    volumes:
+      - db_data_2:/var/lib/postgresql
+    networks:
+      - db
+    #    healthcheck:
+    #      test: ["CMD", "docker-healthcheck-data"]
+    deploy:
+      mode: global
+volumes:
+  db_gtm_1: {}
+  db_coord_1: {}
+  db_coord_2: {}
+  db_data_1: {}
+  db_data_2: {}
+networks:
+  db:
+    driver: overlay
+    attachable: true
```

Comment:
No need for `attachable`, unless you're deploying as a standalone cluster that other Swarm services connect to. If attachable is desired, make sure it's for the Coordinator network only (`db_b`), not the inter-cluster network (`db_a`). These networks are named in order, since in fact because of how Docker Swarm sets up default routes within the container, another order will result in `db_b` being the default for Coordinators, leading to an authentication failure since they're not routing through the `trust` backend network. This can also happen even with Stack deploy, in some cases—safest is in fact to create the networks manually, or to check the networks to ensure that `db_a` has a lower IP subnet than `db_b` (given how the default healthchecks work, which rely on this, so as no longer to require variables passing in the network values).

Reply:
Yes you correct with this. I was just using attachable as I have another container that i was running psql from to test but I definitely wouldn't run it like this production.

Reply:
`attachable` is fine, if this is your use-case. But it's not necessary if you have services launched in the same stack as your Postgres-XL cluster. If you'd rather configure separately, though, or are setting up a single Postgres-XL cluster for multiple services, then `attachable` is fine—or perhaps even publishing ports, if you're sure you're behind a properly-restricted firewall (or perhaps have multiple NICs, even). In the latter case, however, be very careful to only expose the Coordinator-only (`db_b`) overlay network; otherwise, you'll grant public access without any authentication, like noted above for `pg_hba.conf` `trust` vs `md5` (if you're doing this, I highly recommend you test it carefully, too, before going live, in case there's a bug in this program somewhere—no warranty, etc. etc. etc. :) ).